torch/csrc/CudaIPCTypes.cpp

```diff
@@ -0,0 +1,110 @@
+#include <torch/csrc/CudaIPCTypes.h>
+#include <iostream>
+#include <map>
+
+namespace torch {
+
+CudaIPCReceivedData::CudaIPCReceivedData(std::shared_ptr<void> shared_ptr)
+    : shared_ptr_(std::move(shared_ptr)) {}
+
+int64_t CudaIPCSentData::get() {
+  return *counter_ptr;
+}
+
+CudaIPCSentDataLimbo::~CudaIPCSentDataLimbo() {
+  collect();
+  if (shared_blocks_.size() > 0) {
+    std::cerr
+        << "Producer process has been terminated before all shared CUDA tensors released.\n";
+  }
+}
+
+void CudaIPCSentDataLimbo::collect() {
+  std::vector<std::unique_ptr<CudaIPCSentData>> kept_blocks;
+  for (auto& sd : shared_blocks_) {
+    if (sd->get() > 0) {
+      kept_blocks.push_back(std::move(sd));
+    } else {
+      sd.reset();
+    }
+  }
+  shared_blocks_ = std::move(kept_blocks);
+}
+
+CudaIPCSentDataLimbo CudaIPCSentDataLimbo;
+
+CudaIPCSentData::~CudaIPCSentData() {
+  ReturnRefCounter(handle, offset);
+}
+
+void CudaIPCSentDataLimbo::add(std::unique_ptr<CudaIPCSentData> shared_block) {
+  static bool warned = false;
+  if (shared_blocks_.size() > CUDA_IPC_WARN_AFTER_X_BLOCKS_IN_LIMBO &&
+      !warned) {
+    std::cerr
+        << "Producer process tried to deallocate over "
+        << CUDA_IPC_WARN_AFTER_X_BLOCKS_IN_LIMBO
+        << " memory blocks referred by consumer processes. Deallocation might be significantly slowed down.\n";
```

Comment:
This doesn't say what the user should do in the event they see this message. We should probably suggest that they file a bug with PyTorch with the a reproducer (or barring that, some description of how their multiprocessing works.)

Reply:
Hmm... maybe we should use glog for these messages, e.g., using the `LOG(severity)` macro (there are a number of examples lying around.)